### ISSUE TITLE
Support function types (TypeScript), a.k.a. typedefs (Dart)

### ISSUE THREAD

**Issue by yjbanov** [angular/ts2dart#9]
This:

```
interface SearchFunc {
  (source: string, subString: string): boolean;
}
```

Should transpile to this:

```
typedef bool SearchFunc(String source, String subString);
```

**Comment by chalin** [comment]
Are you suggesting that all interfaces defining a single function be interpreted as a typedef? If not, it might be simpler if such interfaces were marked, e.g., `@Typedef`. Thoughts?

**Comment by yjbanov** [comment]
@chalin Note that the `SearchFunc` interface above contains _only_ a signature and _no_ method name. This is TypeScript's way of defining function types, which is very similar to Dart's typedefs.

**Comment by chalin** [comment]
Why not use a function **type alias** (see Section 3.9 of the [TS Spec](http://www.typescriptlang.org/Content/TypeScript%20Language%20Specification.pdf)): e.g.
```ts
type SearchFunc = (source: string, subString: string) => boolean;
```
This matches Dart more closely.

I know that a type alias is different in that an interface *names* a distinct type (though typing is (mostly) structural in TS so I'm not sure that actually is a *significant* difference). Was the use of interfaces like this an idiom in TS used prior to 1.4?

**Comment by yjbanov** [comment]
I actually don't know enough about TypeScript to say. Seems like we should translate both to `typedef`. @mprobst what do you think?

**Comment by chalin** [comment]
I believe that Angular source is being gently massaged into legal TS; IMHO it might be easier / better in the long term to replace the occurrences of the interface-function-type idiom with 1.4 function type alias.

**Comment by chalin** [comment]
[Here is a link](http://www.typescriptlang.org/Playground#src=type%20SearchFunc%20%3D%20(source%3A%20string%2C%20subString%3A%20string)%20%3D%3E%20boolean%3B%0D%0A%0D%0Avar%20f%20%3A%20SearchFunc%20%3D%20(src%2C%20substr)%20%3D%3E%20true%3B) if you want to see this `SearchFunc` example in the TypeScript playgournd.

**Comment by tbosch** [comment]
Same please for inline types:
```
Typescript:
finalize(callback: (CssSelector, any) => void) { ... }

Dart:
finalize(void callback(CssSelector, any)) { ... }
```

Need this for parts in the Angular compiler...

**Comment by mprobst** [comment]
Since 54ef7ba ts2dart at least ignores function types, maybe that helps a bit?

**Comment by mprobst** [comment]
Ping @yjbanov @tbosch

**Comment by yjbanov** [comment]
@mprobst No, I think it makes the situation worse as we pretend that we support them so people can use them (I'm guessing we already do, otherwise why would that commit be there?), but they just disappear in the Dart code.

**Comment by mprobst** [comment]
I don't think I agree with this making the situation worse, I think it's a strict improvement as we can use and provide function types for our TypeScript users.

I don't think I'll get around to doing this soon, maybe you'd be interested in contributing a PR here?

**Comment by yjbanov** [comment]
@mprobst the reason it is worse is because it allows the two codebases to deviate farther from each other than they otherwise could. As an example, currently one can define a Java-style single-method interface as a way to express a function type. It's not ideal, but would work in both languages. However, by allowing typedefs and not supporting them in Dart, people will define them but they will only work in TS. This is worse.

What's on your plate right now? I can talk to @naomiblack so we could prioritize this properly. To fix this issue is now more important given we let https://github.com/angular/ts2dart/commit/54ef7baf77746ce4083140e9a565bba86731f878 to go in.

**Comment by mprobst** [comment]
@yjbanov TypeScript's type system is a super set of Dart's. My judgement is that as much as possible, we should not have TypeScript users (which is both the Angular developers and users of Angular) suffer from being forced into the intersection of Dart and TypeScript; ideally they should be free to use as much of TypeScript's capabilities as possible.

I think supporting Dart's function types would be an improvement, but it's not a high priority for me. If you think it's very important, we should discuss this. As I said, I'd be happy to review a pull request.

**Comment by mprobst** [comment]
`package:func` defines a set of function types that can be parameterized with generic types, might be an easier way to get proper function types.
